src/Func.cpp

```diff
@@ -289,7 +289,8 @@ void Stage::set_dim_type(VarOrRVar var, ForType t) {
             // If it's an rvar and the for type is parallel, we need to
             // validate that this doesn't introduce a race condition.
             if (!dims[i].is_pure() && var.is_rvar && is_parallel(t)) {
-                user_assert(definition.schedule().allow_race_conditions())
+                user_assert(definition.schedule().allow_race_conditions() ||
+                            definition.schedule().atomic())
                     << "In schedule for " << name()
                     << ", marking var " << var.name()
                     << " as parallel or vectorized may introduce a race"
```

Comment:
We should probably augment the assertion-failure text to also mention `.atomic()` as an option.

Reply:
Added something.

Reply:
Maybe I'm confused, but: I was assuming that if you specify `.atomic()`, it makes race conditions impossible. Am I incorrect?

If my assumption is correct, then what we want is something like:

    user_assert(allow_race && !atomic, "existing warning test plus comment about how using .atomic() may prevent the race condition")

Reply:
This is actually a bit complicated: parallel reduction variable can change the meaning of the algorithm even with atomic, which breaks Halide's philosophy. Therefore we want to prove associativity when specifying atomic(). However, as mentioned in the allows_race_condition error message, sometimes the user may still want parallel reduction even when the reduction is not associative (or Halide can't prove it), so we still want something like allow race condition for atomic updates.

Perhaps the problem is the naming of allow race condition: maybe should call it force_parallel_reduction or something like that.

Reply:
My current solution to this is a flag in the atomic function that can turn off the associativity test (see https://github.com/halide/Halide/pull/3936/commits/a26821fa9065d766feedeeaf0f83377963fac6f7).